packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,168 @@
+
+This is achievable because the Nimble `Table` currently tracks when particular rows are collapsed by their row id, which both group and parent rows have. When the Tanstack state has the singleton value of `true` we know we are in a default expand/collapsed state (meaning the user hasn't interactively changed anything), as otherwise it will be a set of id values matched with a boolean state. This will allow us to implement a behavior in `getIsRowExpanded` that will denote group rows as expanded, but parent rows as collapsed, specifically when the TanStack expanded state is set to `true`.
+
+Once _any_ row has been expanded or collapsed, we must update the Nimble `Table` state where we track collapsed rows with _all_ rows that are currently collapsed. This will be a one-time cost. Prototyping suggests that the performance penalty isn't that noticeable.
+
+### Showing a progress indicator for lazy loading
+
+After a user clicks on a parent row to expand it, and that parent does not have children yet, there could be a noticeable delay between when the user clicked the expand button, and when the children show up in the table. This could happen for a variety of reasons, but commonly might occur due to a slow network request.
+
+It would be helpful for the `Table` to visually represent the state that it is currently waiting on data to be loaded from the client. It can accomplish this fairly easily using existing state with no additional APIs or new components.
+
+Prototype visual:
+
+![Lazy Loading Spinner](./spec-images/LazyLoadingSpinner.gif)
+
+This ultimately will put the burden on the client to ensure that the `Table` is updated as needed to get rid of any displayed progress indicator, including in scenarios where the expansion of a parent row failed to load any children (possibly due to some client-side error). The `Table` will only guarantee that the progress indicator is shown when a parent row is expanded and it currently has no children, and that it will be removed once children are present.
+
+Expected user workflow:
+
+1. User loads data into table that specifies some rows as parents, but has no rows that indicate that row as a parent.
+2. User clicks on a parent row
+3. Row expands showing "row loading" indicator
+4. User handles event that row was expanded
+5. User sets data on table that has children for the row that was expanded
+6. "Row loading" indicator is removed and child rows are now displayed
+
+Error workflow:
```

Comment:
Guess we also need to document a cancellation workflow that apps are expected to implement. i.e. if the row is collapsed maybe cancel the request? Or maybe don't cancel it?
Seems like it would be weird to pop open after it was collapsed once data loads.